Fix part of #10415: Refactor classifier validators

rlhouser opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #10415.
2. This PR does the following: Refactors classifier validators to a separate file.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.

Commit:
"refactor classifier validators"

Commit:
refector classifier validators

EricZLou commented:
Big kudos for putting up this PR! One small tip: could you rename this PR to "Refactor classifier validators"? I'll approve your PR once you change that!

EricZLou reviewed (approved):
Looks good to me!

rlhouser commented:
Thank you, and I apologize about the title.

rlhouser commented:
I see I am failing some tests; is it possible to keep working to fix the issues?

EricZLou commented:
Absolutely! You're free to work on this after OSD and there's no rush. Don't worry about the e2e tests if they fail -- just focus on the backend tests. I'm happy to help : )

DubeySandeep commented on core/domain/prod_validators_test.py:
```diff
@@ -146,7 +144,7 @@ def test_model_with_last_updated_greater_than_current_time(self):
         mocked_datetime = datetime.datetime.utcnow() - datetime.timedelta(
             hours=13)
-        with datastore_services.mock_datetime_for_datastore(mocked_datetime):
+        with self.mock_datetime_for_audit(mocked_datetime):
```
Why these changes? is this expected?

rlhouser replied:
I think I have corrected this now.

EricZLou replied:
Ditto on Sandeep's review of this!

DubeySandeep commented on core/domain/prod_validators_test.py:
```diff
@@ -156,7 +154,7 @@ def test_model_with_non_existent_user_id(self):
             'RoleQueryAuditModel\', '
             '[u"Entity id %s: based on field user_ids having value '
-            '%s, expected model UserSettingsModel with '
+            '%s, expect model UserSettingsModel with '
```
Is this change expected?

rlhouser replied:
Updated.

DubeySandeep reviewed (changes requested):
Left a few comments for the test file which I missed earlier, PTAL!

rlhouser commented:
> Absolutely! You're free to work on this after OSD and there's no rush. Don't worry about the e2e tests if they fail -- just focus on the backend tests. I'm happy to help : )

Hello. I am having trouble debugging a failed test related to my changes. I posted in Gitter a couple of days ago but did not get a response. Can you tell me what is the best way to ask for help?

EricZLou commented:
Hey @rlhouser you can also get help by just responding in this PR! Which failed test are you referring to?

rlhouser commented:
Thanks. The test I am checking is core.domain.classifier_validators_test.TrainingJobExplorationMappingModelValidatorTests

When it fails, it gives this message:  

Traceback (most recent call last):
  File "<path>/classifier_validators_test.py", line 257, in test_missing_exploration_model_failure
    expected_output, sort=True, literal_eval=False)
  File "<path>/test_utils.py", line 2806, in run_job_and_check_output
    self.assertEqual(sorted(actual_output), sorted(expected_output))
AssertionError: Lists differ: [u'[u\'failed validation check... != [u'[u\'failed validation check...

First differing element 0:
u'[u\'failed validation check for exploration_ids field check of TrainingJobExplorationMappingModel\', [u"Entity id 0.1.StateTest0: based on field exploration_ids having value 0, expect model ExplorationModel with id 0 but it doesn\'t exist,Entity id 0.1.StateTest0: based on field exploration_ids having value 0, expected model ExplorationModel with id 0 but it doesn\'t exist"]]'
u'[u\'failed validation check for exploration_ids field check of TrainingJobExplorationMappingModel\', [u"Entity id 0.1.StateTest0: based on field exploration_ids having value 0, expect model ExplorationModel with id 0 but it doesn\'t exist"]]'

- [u'[u\'failed validation check for exploration_ids field check of TrainingJobExplorationMappingModel\', [u"Entity id 0.1.StateTest0: based on field exploration_ids having value 0, expect model ExplorationModel with id 0 but it doesn\'t exist,Entity id 0.1.StateTest0: based on field exploration_ids having value 0, expected model ExplorationModel with id 0 but it doesn\'t exist"]]',
? 

I have found where the expected output is defined, but I am not sure how the actual output is generated. I am also not sure how to determine which is correct and if the expected output should be changed or if something else is broken. I would appreciate guidance on debugging. Thanks.

EricZLou commented on core/domain/prod_validators.py:
```diff
@@ -148,243 +145,6 @@ def _get_external_id_relationships(cls, item):
-            cls._validate_state_name]
-
-
-class TrainingJobExplorationMappingModelValidator(
```
Taking a closer look -- you don't need to remove this class from `prod_validators.py`!

rlhouser replied:
When I keep TrainingJobExplorationMappingModelValidator in prod_validators.py, the code fails the core.domain.prod_validators_test.TrainingJobExplorationMappingModelValidatorTests. 

The error messages look something like this:

ERROR:root:Traceback (most recent call last):
  File "/Users/rlhouser/Documents/opensource/oppia/third_party/python_libs/mapreduce/handlers.py", line 503, in handle
    tstate.input_reader, shard_state, tstate, ctx)
  File "/Users/rlhouser/Documents/opensource/oppia/third_party/python_libs/mapreduce/handlers.py", line 593, in _process_inputs
    entity, input_reader, ctx, tstate):
  File "/Users/rlhouser/Documents/opensource/oppia/third_party/python_libs/mapreduce/handlers.py", line 635, in _process_datum
    for output in result:
  File "/Users/rlhouser/Documents/opensource/oppia/core/domain/prod_validation_jobs_one_off.py", line 120, in map
    validator = getattr(validator_module, validator_cls_name)
AttributeError: 'module' object has no attribute 'TrainingJobExplorationMappingModelValidator'

As best as I can tell, this has something to do with the fact that in prod_validation_jobs_one_off.py, the method TrainingJobExplorationMappingModelAuditOneOffJob.entity_classes_to_map_over returns the classifier_models.TrainingJobExplorationMappingModel. Can you tell me if it seems I need to change something in core/domain/prod_validation_jobs_one_off.py?  Or possibly in core/storage/classfier/gae_models.py?

EricZLou commented on core/domain/prod_validators.py:
```diff
@@ -23,10 +23,7 @@
 import itertools
 import re
 
-from constants import constants
```
Keep `from constants import constants` since we're keeping `TrainingJobExplorationMappingModelValidator` in this file!

EricZLou commented on core/domain/prod_validators.py:
```diff
@@ -23,10 +23,7 @@
 
-from constants import constants
 from core.domain import base_model_validators
-from core.domain import classifier_domain
```
Keep `from core.domain import classifier_domain` since we're keeping `TrainingJobExplorationMappingModelValidator` in this file!

EricZLou commented on core/domain/prod_validators.py:
```diff
@@ -689,7 +449,7 @@ def _validate_node_count(
                         base_model_validators.ERROR_CATEGORY_FIELD_CHECK),
                     'Entity id %s: based on field collection_ids having'
-                    ' value %s, expected model %s with id %s but it doesn\'t'
+                    ' value %s, expect model %s with id %s but it doesn\'t'
```
Why'd we make these changes? Maybe we can undo them? (including all of the similar changes below)

EricZLou commented on core/domain/prod_validators.py:
```diff
@@ -2873,182 +2633,6 @@ def _get_model_domain_object_instance(cls, item):
             .create_community_contribution_stats_from_model(item)
         )
 
-    @classmethod
```
Why did we remove these functions (`_validate_translation_reviewer_counts`, `_validate_translation_suggestion_counts`, etc.)? Maybe we should put them back.

EricZLou commented on core/domain/prod_validators_test.py:
```diff
@@ -273,265 +271,6 @@ def test_model_with_invalid_id(self):
-            expected_output, sort=True, literal_eval=False)
-
-
-class TrainingJobExplorationMappingModelValidatorTests(
```
We can keep this test class here since we're keeping `TrainingJobExplorationMappingModelValidator` in `prod_validators.py`.

EricZLou commented on core/domain/prod_validators_test.py:
```diff
@@ -9231,20 +8916,11 @@ def test_model_validation_fails_for_negative_translation_reviewer_counts(
             'fails domain validation with the error Expected the translation '
             'reviewer count to be non-negative for %s language code, '
-            'received: %s.\']]' % (
+            'recieved: %s.\']]' % (
```
Do you remember making this change here from "received" to "recieved"?

rlhouser replied:
No; I don't remember changing that. I am pretty sure only cut, copied, and pasted.

EricZLou reviewed (changes requested):
Left some more comments! Please take a look and let me know if there are parts of the code that you don't remember changing.

Commit:
refector classifier validators

Commit:
refactor classifier validators

Commit:
refactor classifier validators

EricZLou commented:
Thanks for continuing to work on this @rlhouser go go go!!! Let me know how I can help -- you can chat me at ericzlou@gmail.com

rlhouser commented:
A question regarding one of the comments: when I keep TrainingJobExplorationMappingModelValidator in prod_validators.py, the code fails the core.domain.prod_validators_test.TrainingJobExplorationMappingModelValidatorTests.

The error messages look something like this:

ERROR:root:Traceback (most recent call last):
File "/Users/rlhouser/Documents/opensource/oppia/third_party/python_libs/mapreduce/handlers.py", line 503, in handle
tstate.input_reader, shard_state, tstate, ctx)
File "/Users/rlhouser/Documents/opensource/oppia/third_party/python_libs/mapreduce/handlers.py", line 593, in _process_inputs
entity, input_reader, ctx, tstate):
File "/Users/rlhouser/Documents/opensource/oppia/third_party/python_libs/mapreduce/handlers.py", line 635, in _process_datum
for output in result:
File "/Users/rlhouser/Documents/opensource/oppia/core/domain/prod_validation_jobs_one_off.py", line 120, in map
validator = getattr(validator_module, validator_cls_name)
AttributeError: 'module' object has no attribute 'TrainingJobExplorationMappingModelValidator'

As best as I can tell, this has something to do with the fact that in prod_validation_jobs_one_off.py, the method TrainingJobExplorationMappingModelAuditOneOffJob.entity_classes_to_map_over returns the classifier_models.TrainingJobExplorationMappingModel. Can you tell me if it seems I need to change something in core/domain/prod_validation_jobs_one_off.py? Or possibly in core/storage/classfier/gae_models.py?

EricZLou commented:
Hi Rebekah,

Just to be sure we're on the same page, you've only made these changes to the following files:
- **classifier_validators.py:** add `ClassifierTrainingJobModelValidator`
- **classifier_validators_test.py:** add `ClassifierTrainingJobModelValidatorTests`
- **prod_validators.py:** remove `ClassifierTrainingJobModelValidator`
- **prod_validators_test.py:** remove `ClassifierTrainingJobModelValidatorTests`

rlhouser commented:
Yes, I only changed those 4 files.

EricZLou commented:
Can you make sure that you haven't deleted `TrainingJobExplorationMappingModelValidator` from `prod_validators.py` and that you haven't deleted `TrainingJobExplorationMappingModelValidatorTests` from `prod_validators_test.py`? I don't think we should need to change anything in `prod_validation_jobs_one_off.py` or in `gae_models.py`. Thanks!

rlhouser commented:
In the tests I am running/failing, TrainingJobExplorationMappingModelValidator  is in prod_validators.py and TrainingJobExplorationMappingModelValidatorTests is in  prod_validators_test.py . Just to clarify: in my last PR, I had removed TrainingJobExplorationMappingModelValidator from prod_validators.py. and put it in classifier_validators.py. I realized I was supposed to keep it in prod_validators.py, so I moved it back. Thanks.

EricZLou commented:
`TrainingJobExplorationMappingModelValidatorTests` shouldn't fail if they're in `prod_validators_test.py` since they were originally in that file. I've sent over some advice through email!

EricZLou commented:
Commenting to un-stale this issue!

rlhouser commented:
> Commenting to un-stale this issue!

Thank you.

rlhouser commented:
I have opened a new PR for this issue: #11116.
